lib/active_merchant/billing/gateways/blue_snap.rb

```diff
@@ -204,6 +205,46 @@ def add_address(doc, options)
         doc.zip(address[:zip]) if address[:zip]
       end
 
+      def add_level_3_data(doc, options)
+        doc.send('level-3-data') do
+          send_when_present(doc, :customer_reference_number, options)
+          send_when_present(doc, :sales_tax_amount, options)
+          send_when_present(doc, :freight_amount, options)
+          send_when_present(doc, :duty_amount, options)
+          send_when_present(doc, :destination_zip_code, options)
+          send_when_present(doc, :destination_country_code, options)
+          send_when_present(doc, :ship_from_zip_code, options)
+          send_when_present(doc, :discount_amount, options)
+          send_when_present(doc, :tax_amount, options)
+          send_when_present(doc, :tax_rate, options)
+          add_level_3_data_items(doc, options[:level_3_data_items]) if options[:level_3_data_items]
+        end
+      end
+
+      def send_when_present(dest, dest_path, source, source_path = nil)
```

Comment:
I suggest renaming `dest` to `doc` to be consistent. I also recommended renaming `dest_path` to `options_key` or similar, `source` to `options` and `source_path` to `xml_element_name` (or equivalent - `element_name`, `elem_name`).

Couldn't lines 225 - 234 be  condensed to: 

```
source_path ||= dest_path
return unless source[key]

dest_path = dest_path.to_s.dasherize
dest.send(dest_path, source)
```

Reply:
Condensing it that way (or a few other ways) doesn't seem to work. I adopted this pattern from something very similar, implemented on the Stripe gateway. Updated to re-name variables for better clarity/ consistency.

Reply:
@molbrown try 

```
def send_when_present(doc, options_key, options, xml_element_name)
    xml_element_name ||= options_key
    return unless options[options_key]

    xml_element_name = xml_element_name.to_s.dasherize
    doc.send(xml_element_name, options[options_key])
end
```

Reply:
Nice! Updated.

Reply:
💥